src/dataframerow/dataframerow.jl

```diff
@@ -259,11 +259,45 @@ for T in MULTICOLUMNINDEX_TUPLE
     end
 end
 
-Base.@propagate_inbounds Base.setindex!(r::DataFrameRow, value, idx) =
-    setindex!(parent(r), value, row(r), parentcols(index(r), idx))
-
 index(r::DataFrameRow) = getfield(r, :colindex)
 
+is_column_insertion_allowed(dfr::DataFrameRow) = index(dfr) isa Index
+
+Base.@propagate_inbounds function Base.setindex!(dfr::DataFrameRow, value, idx)
+    if idx isa SymbolOrString && columnindex(dfr, idx) == 0
+        if !is_column_insertion_allowed(dfr)
+            throw(ArgumentError("creating new columns in a DataFrameRow that subsets " *
+                                "columns of its parent data frame is disallowed"))
+        end
+        T = typeof(val)
+        newcol = similar(val, Union{T,Missing}, nrow(parent(dfr)))
```

Comment:
This is a very good point. I used `similar`, because we already had it in
https://github.com/JuliaData/DataFrames.jl/blob/main/src/subdataframe/subdataframe.jl#L207
which also requires fixing then.

Other places with the same issue are e.g.:
* https://github.com/JuliaData/DataFrames.jl/blob/8c32d537b1d5a2ae9b1cdd72928def18aeae0bf9/src/subdataframe/subdataframe.jl#L239
* https://github.com/JuliaData/DataFrames.jl/blob/8c32d537b1d5a2ae9b1cdd72928def18aeae0bf9/src/other/broadcasting.jl#L204
* https://github.com/JuliaData/DataFrames.jl/blob/8c32d537b1d5a2ae9b1cdd72928def18aeae0bf9/src/dataframe/insertion.jl#L313

Do you think we should do a systematic review of them and fix this everywhere?

Reply:
I thought we followed the policy that `similar` is called when we have a vector to pass to it, and otherwise `allocatecolumn` is used. Apparently that's not really the case, given the example in broadcasting.jl. The two other examples look correct to me. We could try changing the broadcasting.jl one and check whether other places have the same issue.